csrc/multidevice/communicator.cpp

```diff
@@ -158,32 +160,54 @@ Communicator::Communicator(
                            master_addr_ == gethostbyname(hostname)->h_name) &&
         local_rank_ == server_local_rank;
   }
-  if (master_port_) {
-    store_opts.port = master_port_;
+  store_opts.port = master_port_ ? master_port_ : comm_master_port_default;
+  store_ = c10::make_intrusive<c10d::TCPStore>(master_addr_, store_opts);
+
+  // creates the world's backend
+  std::vector<RankType> all_ranks(size_);
+  std::iota(all_ranks.begin(), all_ranks.end(), 0);
+  world_ = getBackendForTeam(all_ranks);
+}
+
+c10::intrusive_ptr<c10d::Backend> Communicator::getBackendForTeam(
+    const std::vector<DeviceIdxType>& ranks) {
+  // check if backend associated with the ranks is present in the cache
+  if (teams_.find(ranks) == teams_.end()) { // create the backend and cache it
+    // check that the caller's rank belongs to the requested team
+    auto rank_it = std::find(ranks.begin(), ranks.end(), rank_);
+    NVF_ERROR(
+        rank_it != ranks.end(),
+        "only ranks in the team should participate to its initialization");
+    // retrieve the caller's rank index/position in the team
+    RankType team_rank = std::distance(ranks.begin(), rank_it);
+    // generate a string key which is unique to the team
+    std::string team_key = std::accumulate(
```

Comment:
Should we make `teams_` an unordered map from `team_key` to backends? This way, we do not need the `O(logN)` time for `std::map` lookup.

Reply:
done!